File: 22-Strategy/src/go/shopping_cart.go
package main

import "fmt"

// ShoppingCart is the Context — it holds a payment strategy and delegates
// checkout to it. The strategy can be swapped at runtime.
type ShoppingCart struct {
	total    float64
	strategy PaymentStrategy
}

func (s *ShoppingCart) AddItem(name string, price float64) {
	s.total += price
	fmt.Printf("  + %s ($%.2f)\n", name, price)
}

func (s *ShoppingCart) SetPaymentStrategy(strategy PaymentStrategy) {
	s.strategy = strategy
}

func (s *ShoppingCart) Checkout() {
	if s.strategy == nil {
		fmt.Println("  No payment method selected!")
		return
	}
	s.strategy.Pay(s.total)
	s.total = 0
}

File: 22-Strategy/src/rust/shopping_cart.rs
use crate::payment_strategy::PaymentStrategy;

/// Context — holds a payment strategy and delegates checkout to it. The
/// strategy (a boxed trait object) can be swapped at runtime.
pub struct ShoppingCart {
    total: f64,
    strategy: Option<Box<dyn PaymentStrategy>>,
}

impl ShoppingCart {
    pub fn new() -> Self {
        ShoppingCart { total: 0.0, strategy: None }
    }

    pub fn add_item(&mut self, name: &str, price: f64) {
        self.total += price;
        println!("  + {} (${:.2})", name, price);
    }

    pub fn set_payment_strategy(&mut self, strategy: Box<dyn PaymentStrategy>) {
        self.strategy = Some(strategy);
    }

    pub fn checkout(&mut self) {
        match &self.strategy {
            Some(strategy) => {
                strategy.pay(self.total);
                self.total = 0.0;
            }
            None => println!("  No payment method selected!"),
        }
    }
}
